File: mobile/android/app/src/test/java/app/terra/explore/TrackingTests.kt
package app.terra.explore
import org.junit.Test
import org.junit.Assert.*
class TrackingTests {
    @Test fun overlappingDiscoveryCountedOnce() {
        val s=Session(mode=Mode.walk,startedAt=0,points=mutableListOf(Point(55.0,37.0,1000,5.0)))
        assertTrue(Discovery.cells(listOf(s)).isNotEmpty()); assertEquals(Discovery.cells(listOf(s)),Discovery.cells(listOf(s,s)))
    }
    @Test fun filtersGPS() {
        val a=Point(55.0,37.0,1000,5.0)
        assertFalse(Point(55.0,37.1,2000,5.0).accepts(a,Mode.walk))
        assertFalse(Point(55.0,37.001,10000,100.0).accepts(a,Mode.walk))
        assertTrue(Point(55.0,37.0001,11000,5.0).accepts(a,Mode.walk))
    }
    @Test fun pauseAndGPSGap() {
        val s=Session(mode=Mode.walk,startedAt=1000,endedAt=110000,pausedMs=10000,points=mutableListOf(Point(55.0,37.0,1000,5.0),Point(55.0,37.001,11000,5.0,true),Point(55.0,37.002,100000,5.0)))
        assertEquals(0.0,s.distance(),0.01); assertEquals(99L,s.duration())
    }
    @Test fun backupRoundTrip() {
        val s=Session(mode=Mode.bike,startedAt=1000,endedAt=3000,points=mutableListOf(Point(55.0,37.0,2000,5.0)))
        val decoded=Session.parse(s.json()); decoded.validate(); assertEquals(s.id,decoded.id); assertEquals(s.points,decoded.points)
    }
    @Test(expected=IllegalArgumentException::class) fun rejectsMalformedBackup() { Session(mode=Mode.walk,startedAt=1000,endedAt=500).validate() }
}

File: mobile/ios/TerraTests/TrackingTests.swift
import XCTest
@testable import Terra
final class TrackingTests: XCTestCase {
    func testRejectsBadGPSAndSpeedSpikes() {
        let a = TrackPoint(lat: 55, lng: 37, t: 1000, accuracy: 5)
        XCTAssertFalse(TrackPoint(lat: 55, lng: 37.1, t: 2000, accuracy: 5).accepts(after: a, mode: .walk))
        XCTAssertFalse(TrackPoint(lat: 55, lng: 37.001, t: 10000, accuracy: 100).accepts(after: a, mode: .walk))
        XCTAssertTrue(TrackPoint(lat: 55, lng: 37.0001, t: 11000, accuracy: 5).accepts(after: a, mode: .walk))
    }
    func testPauseAndSignalLossDoNotConnect() {
        let a = TrackPoint(lat: 55, lng: 37, t: 1000, accuracy: 5)
        let b = TrackPoint(lat: 55, lng: 37.001, t: 11000, accuracy: 5, breakBefore: true)
        let c = TrackPoint(lat: 55, lng: 37.002, t: 100000, accuracy: 5)
        let s = TrackSession(mode: .walk, startedAt: 1000, endedAt: 110000, pausedMs: 10000, points: [a,b,c])
        XCTAssertEqual(s.distance, 0); XCTAssertEqual(s.duration(), 99)
    }
    func testBackupRoundTripAndValidation() throws {
        let s = TrackSession(mode: .bike, startedAt: 1000, endedAt: 3000, points: [TrackPoint(lat: 55, lng: 37, t: 2000, accuracy: 5)])
        let decoded = try JSONDecoder().decode(Backup.self, from: JSONEncoder().encode(Backup(sessions: [s])))
        try decoded.validate(); XCTAssertEqual(decoded.sessions[0].id, s.id)
        XCTAssertThrowsError(try Backup(sessions: [s,s]).validate())
        XCTAssertThrowsError(try TrackSession(mode: .walk, startedAt: 1000, endedAt: 500).validate())
    }
    func testDiscoveryDoesNotDoubleCountOverlap() {
        let s = TrackSession(mode: .walk, startedAt: 0, points: [TrackPoint(lat: 55, lng: 37, t: 1000, accuracy: 5)])
        XCTAssertFalse(Discovery.cells([s]).isEmpty)
        XCTAssertEqual(Discovery.cells([s]), Discovery.cells([s,s]))
    }
}
